service/firewall/interception/dnslistener/etwlink_windows.go

```diff
@@ -0,0 +1,138 @@
+package dnslistener
+
+import (
+	"fmt"
+	"runtime"
+	"sync"
+	"sync/atomic"
+
+	"golang.org/x/sys/windows"
+)
+
+type ETWSession struct {
+	dll *windows.DLL
+
+	createState       *windows.Proc
+	initializeSession *windows.Proc
+	startTrace        *windows.Proc
+	flushTrace        *windows.Proc
+	stopTrace         *windows.Proc
+	destroySession    *windows.Proc
+
+	shutdownGuard atomic.Bool
+	shutdownMutex sync.Mutex
+
+	state uintptr
+}
+
+// NewSession creates new ETW event listener and initilizes it. This is a low level interface, make sure to call DestorySession when you are done using it.
+func NewSession(dllpath string, callback func(domain string, result string)) (*ETWSession, error) {
+	etwListener := &ETWSession{}
+
+	// Initialize dll functions
+	var err error
+	etwListener.dll, err = windows.LoadDLL(dllpath)
+	if err != nil {
+		return nil, fmt.Errorf("faild to load dll: %q", err)
+	}
+	etwListener.createState, err = etwListener.dll.FindProc("PM_ETWCreateState")
+	if err != nil {
+		return nil, fmt.Errorf("faild to load function PM_ETWCreateState: %q", err)
+	}
+	etwListener.initializeSession, err = etwListener.dll.FindProc("PM_ETWInitializeSession")
+	if err != nil {
+		return nil, fmt.Errorf("faild to load function PM_ETWInitializeSession: %q", err)
+	}
+	etwListener.startTrace, err = etwListener.dll.FindProc("PM_ETWStartTrace")
+	if err != nil {
+		return nil, fmt.Errorf("faild to load function PM_ETWStartTrace: %q", err)
+	}
+	etwListener.flushTrace, err = etwListener.dll.FindProc("PM_ETWFlushTrace")
+	if err != nil {
+		return nil, fmt.Errorf("faild to load function PM_ETWFlushTrace: %q", err)
+	}
+	etwListener.stopTrace, err = etwListener.dll.FindProc("PM_ETWStopTrace")
+	if err != nil {
+		return nil, fmt.Errorf("faild to load function PM_ETWStopTrace: %q", err)
+	}
+	etwListener.destroySession, err = etwListener.dll.FindProc("PM_ETWDestroySession")
+	if err != nil {
+		return nil, fmt.Errorf("faild to load function PM_ETWDestroySession: %q", err)
+	}
+
+	// Initialize notification activated callback
+	win32Callback := windows.NewCallback(func(domain *uint16, result *uint16) uintptr {
+		callback(windows.UTF16PtrToString(domain), windows.UTF16PtrToString(result))
+		return 0
+	})
+	// The function only allocates memory it will not fail.
+	etwListener.state, _, _ = etwListener.createState.Call(win32Callback)
+
+	// Make sure DestroySession is called even if caller forgets to call it.
+	runtime.SetFinalizer(etwListener, func(l *ETWSession) {
+		_ = l.DestroySession()
+	})
```

Comment:
Im calling DestroySession explicitly. This for resilience, if for some reason the DestorySession is not called because of panic or error. Do you think this is still a bad practice?